File: frontend/src/services/classics.ts
import api from '../utils/axios';
import { Classic, Translation } from '../types/classic';

interface GetClassicsParams {
    skip?: number;
    limit?: number;
    category?: string;
    dynasty?: string;
    tag?: string;
}

interface PaginatedResponse {
    items: Classic[];
    total: number;
    skip: number;
    limit: number;
}

export const getClassics = async (params: GetClassicsParams = {}): Promise<PaginatedResponse> => {
    try {
        console.log('Fetching classics with params:', params);
        const response = await api.get<PaginatedResponse>('/classics', {
            params: {
                skip: params.skip || 0,
                limit: params.limit || 9,
                category: params.category || undefined,
                dynasty: params.dynasty || undefined,
                tag: params.tag || undefined
            }
        });
        console.log('Response data:', response.data);
        return response.data;
    } catch (error) {
        console.error('Error fetching classics:', error);
        throw new Error('获取古籍列表失败');
    }
};

export const getClassic = async (id: number): Promise<Classic> => {
    try {
        const response = await api.get<Classic>(`/classics/${id}`);
        return response.data;
    } catch (error) {
        console.error('Error fetching classic:', error);
        throw error;
    }
};

export const getTranslations = async (classicId: number): Promise<Translation[]> => {
    try {
        const response = await api.get<Translation[]>(`/classics/${classicId}/translations`);
        return response.data;
    } catch (error) {
        console.error('Error fetching translations:', error);
        throw error;
    }
};

export const getClassicById = async (id: number): Promise<Classic> => {
    try {
        const response = await api.get<Classic>(`/classics/${id}`);
        return response.data;
    } catch (error) {
        console.error('Error fetching classic:', error);
        throw new Error('获取古籍详情失败');
    }
};

export const likeClassic = async (id: string): Promise<void> => {
    try {
        await api.post(`/classics/${id}/like`);
    } catch (error) {
        console.error('Error liking classic:', error);
        throw error;
    }
};


export interface SearchClassicsParams {
    query: string;
    searchType?: 'title' | 'content' | 'author' | 'all';
}

export const searchClassics = async (params: SearchClassicsParams) => {
    try {
        console.log('搜索参数:', params);

        // 改用POST请求，将参数放在请求体中
        const response = await api.post('/classics/search', {
            query: params.query,
            search_type: params.searchType || 'all'
        });

        console.log('搜索响应:', response.data);
        return response.data;
    } catch (error: any) {
        console.error('搜索古籍出错:', error);
        if (error.response) {
            console.error('错误响应数据:', error.response.data);
            console.error('错误状态码:', error.response.status);
        }
        throw error;
    }
};

File: backend/routes/classics.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from database import get_db
from models.classic import Classic, Tag, Translation
from schemas.classic import (
    ClassicCreate,
    ClassicUpdate,
    ClassicResponse,
    TagCreate,
    TranslationCreate,
)
from datetime import datetime
from pydantic import BaseModel

router = APIRouter()


class PaginatedResponse(BaseModel):
    items: List[ClassicResponse]
    total: int
    skip: int
    limit: int


@router.get("/classics", response_model=PaginatedResponse)
def get_classics(
    skip: int = 0,
    limit: int = 9,
    category: Optional[str] = None,
    dynasty: Optional[str] = None,
    tag: Optional[str] = None,
    db: Session = Depends(get_db),
):
    """获取古籍列表，支持分页和筛选"""
    # 构建查询条件
    query = db.query(Classic)

    # 应用筛选条件
    if category:
        query = query.filter(Classic.category == category)
    if dynasty:
        query = query.filter(Classic.dynasty == dynasty)
    if tag:
        query = query.join(Classic.tags).filter(Tag.name == tag)

    # 获取总数
    total = query.count()

    # 应用分页
    classics = query.offset(skip).limit(limit).all()

    # 返回分页响应
    return {"items": classics, "total": total, "skip": skip, "limit": limit}


@router.post("/classics", response_model=ClassicResponse)
def create_classic(classic: ClassicCreate, db: Session = Depends(get_db)):
    db_classic = Classic(
        title=classic.title,
        content=classic.content,
        author=classic.author,
        dynasty=classic.dynasty,
        category=classic.category,
        source=classic.source,
    )

    # 处理标签
    if classic.tags:
        for tag_name in classic.tags:
            tag = db.query(Tag).filter(Tag.name == tag_name).first()
            if not tag:
                tag = Tag(name=tag_name)
                db.add(tag)
            db_classic.tags.append(tag)

    db.add(db_classic)
    db.commit()
    db.refresh(db_classic)
    return db_classic


@router.get("/classics/{classic_id}", response_model=ClassicResponse)
def get_classic(classic_id: int, db: Session = Depends(get_db)):
    classic = db.query(Classic).filter(Classic.id == classic_id).first()
    if not classic:
        raise HTTPException(status_code=404, detail="Classic not found")
    return classic


@router.put("/classics/{classic_id}", response_model=ClassicResponse)
def update_classic(
    classic_id: int, classic: ClassicUpdate, db: Session = Depends(get_db)
):
    db_classic = db.query(Classic).filter(Classic.id == classic_id).first()
    if not db_classic:
        raise HTTPException(status_code=404, detail="Classic not found")

    for field, value in classic.dict(exclude_unset=True).items():
        if field == "tags":
            # 更新标签
            db_classic.tags = []
            for tag_name in value:
                tag = db.query(Tag).filter(Tag.name == tag_name).first()
                if not tag:
                    tag = Tag(name=tag_name)
                    db.add(tag)
                db_classic.tags.append(tag)
        else:
            setattr(db_classic, field, value)

    db.commit()
    db.refresh(db_classic)
    return db_classic


@router.delete("/classics/{classic_id}")
def delete_classic(classic_id: int, db: Session = Depends(get_db)):
    classic = db.query(Classic).filter(Classic.id == classic_id).first()
    if not classic:
        raise HTTPException(status_code=404, detail="Classic not found")

    db.delete(classic)
    db.commit()
    return {"message": "Classic deleted successfully"}


@router.post("/classics/{classic_id}/translations", response_model=Translation)
def add_translation(
    classic_id: int, translation: TranslationCreate, db: Session = Depends(get_db)
):
    classic = db.query(Classic).filter(Classic.id == classic_id).first()
    if not classic:
        raise HTTPException(status_code=404, detail="Classic not found")

    db_translation = Translation(
        classic_id=classic_id,
        content=translation.content,
        translator=translation.translator,
        language=translation.language,
    )

    db.add(db_translation)
    db.commit()
    db.refresh(db_translation)
    return db_translation
